### DISCUSSION TITLE
Request for PowerShell to do something about the modules assembly conflict situation

### DISCUSSION THREAD

**Discussion by o-l-a-v** [PowerShell/PowerShell#24365] [Enhancements]
PowerShell should educate and push C# compiled PowerShell module creators and maintainers to follow best practices, like using AssemblyLoadContext to avoid assembly conflicts. More context: <https://youtu.be/__J7b84BAmQ>.

This is a PowerShell problem, it has a huge negative impact on using PowerShell itself. It's not just a problem with an individual module.

To get an indication on how widespread this is: Numbers of issues opened containing "could not load file or assembly" as of 2024-10-03:

| What                       | Count | Url                                                                                                                  |
| :------------------------- | :---- | :------------------------------------------------------------------------------------------------------------------- |
| `Az`                       | 74    | <https://github.com/Azure/azure-powershell/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>                |
| `Microsoft.Graph`          | 34    | <https://github.com/microsoftgraph/msgraph-sdk-powershell/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22> |
| `PnP.PowerShell`           | 33    | <https://github.com/pnp/powershell/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>                        |
| `PSScriptAnalyzer`         | 3     | <https://github.com/PowerShell/PSScriptAnalyzer/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>           |
| PowerShell                 | 134   | <https://github.com/PowerShell/PowerShell/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>                 |
| `PowerShellEditorServices` | 4     | <https://github.com/PowerShell/PowerShellEditorServices/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>   |
| `vscode-powershell`        | 25    | <https://github.com/PowerShell/vscode-powershell/issues?q=is%3Aissue+%22could+not+load+file+or+assembly%22>          |

VSCode with the `vscode-powershell` extension is the official way to author PowerShell now. `vscode-powershell` also uses assembly dependecies that can conflict with assemblies used by modules, which can create issues that does not happen in a PowerShell terminal outside of `vscode-powershell`.

This is not an issue `vscode-powershell` can fix from their context, like this issue (and **many** others like it) describes:

* <https://github.com/PowerShell/vscode-powershell/issues/4727>

A centralized initiative with a more forced approach should be considered IMO. I think the PowerShell project itself should do something about it. Some ideas:

* Require Microsoft owned modules in the PowerShell Gallery to use AssemblyLoadContext for **all** assembly dependencies for **all** of their submodules.
  * Multiple Microsoft modules are still not using AssemblyLoadContext for all their depdendencies, like `Az`, `ExchangeOnlineManagement`, `Microsoft.Graph` and `PnP.PowerShell`.
  * I think breaking Windows PowerShell 5.1 compatibility for fixing this is worth it.
    * Arguably, the decision to break Windows PowerShell 5.1 compatibility was made when using any common assembly as a dependency.
  * Windows PowerShell 5.1 compatibility is already broken with no commitment to fix it in `Az` and `vscode-powershell`.
* Write or recommend a set of tests that all compiled C# modules should pass.

**Comment by rhubarb-geek-nz** [comment]
On the positive side, it is easier to support AssemblyLoadContext in PowerShell 7 than in 5. First a developer has to be convinced to leave `netstandard2.0` behind first. I have published a few with ALC including those with native libraries for multiple platforms, so there are examples around of it being done (hopefully well).

Currently the only gate-keeper of the PSGallery is [Publish-Module](https://learn.microsoft.com/en-us/powershell/module/powershellget/publish-module) itself which does perform validation appropriate for PowerShell, however it is not compulsory.

I don't see the Gallery becoming a gated community, I think that would be the slippery-slope if there were additional hurdles for developers to publish packages, at least they don't (currently) need code signing certificates.

I don't see `vscode-powershell` as being the only path to publishing modules. I happily use full-fat Visual Studio.
